app/controllers/pages_controller.rb

```diff
@@ -10,4 +12,25 @@ def help
       format.html
     end
   end
+
+  private
+
+  def check_uprn_link
+    return if params[:uprn].nil?
+
+    building = Building.find_by(uprn: params[:uprn])
+
+    if building.nil?
+      redirect_to root_path, alert: I18n.t("errors.unknown_uprn", uprn: params[:uprn])
+      return
+    end
+
+    @survey = Survey::Record.find_or_create_by(
```

Comment:
We can create the record through the building and avoid having to set `building_id`:

``` ruby
begin
  survey = building.find_or_create_by!(session_id: session.id)
rescue ActiveRecord::RecordNotUnique
  retry
end

redirect_to goto_url("ownership")
```

The `rescue` is to handle any race conditions.